### PR TITLE
Add RHEL support to Java feature

### LINKED ISSUE #820: Add RHEL support to java feature.
Add RHEL support to java feature. Tests for alma, rocky, fedora and mariner linux included.

PR to address this issue is forthcoming.

### PR DESCRIPTION
**Pull request by jdputschadi**
This PR adds RHEL support to java feature.

CentOS 7, RHEL/Alma/Rocky Linux 8 & 9, Mariner, and Fedor should all work. Tests for all of these are included.

resolves #820

### CODE REVIEW

File: src/java/install.sh [hunk 1]
```diff
@@ -61,31 +112,77 @@ elif [ "${USERNAME}" = "none" ] || ! id -u ${USERNAME} > /dev/null 2>&1; then
...
 }
 
-apt_get_update()
-{
-    if [ "$(find /var/lib/apt/lists/* | wc -l)" = "0" ]; then
-        echo "Running apt-get update..."
-        apt-get update -y
-    fi
+
+pkg_mgr_update() {
```
Comment (samruddhikhandale) [inline src/java/install.sh:139]:
```suggestion
pkg_manager_update() {
```

Suggestion for refactoring the function name!

File: src/java/install.sh [hunk 2]
```diff
@@ -30,14 +30,65 @@ ADDITIONAL_VERSIONS="${ADDITIONALVERSIONS:-""}"
...
+elif type dnf > /dev/null 2>&1; then
+    PKG_MGR_CMD=dnf
+    INSTALL_CMD="${PKG_MGR_CMD} -y install"
+elif type yum > /dev/null 2>&1; then
+    PKG_MGR_CMD=yum
+    INSTALL_CMD="${PKG_MGR_CMD} -y install"
+else
+    echo "(Error) Unable to find a supported package manager."
+    exit 1
+fi
```
Comment (samruddhikhandale) [inline src/java/install.sh:72]: Neat 🎉

File: src/java/install.sh [hunk 3]
```diff
@@ -61,31 +112,77 @@ elif [ "${USERNAME}" = "none" ] || ! id -u ${USERNAME} > /dev/null 2>&1; then
...
+            else
+                if [ "$(ls /var/cache/${PKG_MGR_CMD}/* 2>/dev/null | wc -l)" = 0 ]; then
+                    echo "Running ${PKG_MGR_CMD} check-update ..."
+                    set +e
+                    ${PKG_MGR_CMD} check-update
+                    rc=$?
+                    # centos 7 sometimes returns a status of 100 when it apears to work.
+                    if [ $rc != 0 ] && [ $rc != 100 ]; then
+                        exit 1
+                    fi
```
Comment (samruddhikhandale) [inline src/java/install.sh:162]: Can we indent with a single tab (the lines in between `set`)? This way, I feel like the code is more readable.

File: src/java/install.sh [hunk 4]
```diff
@@ -61,31 +112,77 @@ elif [ "${USERNAME}" = "none" ] || ! id -u ${USERNAME} > /dev/null 2>&1; then
...
+                fi
+            else
+                if [ "$(ls /var/cache/${PKG_MGR_CMD}/* 2>/dev/null | wc -l)" = 0 ]; then
+                    echo "Running ${PKG_MGR_CMD} check-update ..."
+                    set +e
+                    ${PKG_MGR_CMD} check-update
+                    rc=$?
+                    # centos 7 sometimes returns a status of 100 when it apears to work.
+                    if [ $rc != 0 ] && [ $rc != 100 ]; then
+                        exit 1
```
Comment (samruddhikhandale) [inline src/java/install.sh:161]: Can we add an error message on what failed and print `rc` as well?

File: src/java/install.sh [hunk 5]
```diff
@@ -142,8 +239,19 @@ if [ "${architecture}" != "amd64" ] && [ "${architecture}" != "x86_64" ] && [ "$
     exit 1
 fi
 
-# Install dependencies
-check_packages curl ca-certificates zip unzip sed
+# Install dependencies,
+check_packages ca-certificates zip unzip sed findutils util-linux tar
```
Comment (samruddhikhandale) [inline src/java/install.sh:243]: > findutils util-linux tar

Curious, why do we need to additionally install ^ ?

File: src/java/install.sh [hunk 6]
```diff
@@ -142,8 +239,19 @@ if [ "${architecture}" != "amd64" ] && [ "${architecture}" != "x86_64" ] && [ "$
...
+# Make sure passwd (Debian) and shadow-utils RHEL family is installed
```
Comment (samruddhikhandale) [inline src/java/install.sh:244]: Same question here, were we seeing any failures or is it like a nice to have in package?

File: test/java/scenarios.json [hunk 7]
```diff
@@ -69,5 +69,142 @@
                 "jdkDistro": "graalce"
             }
         }
+    },
+    "install_from_non_default_distro_rhel_family": {
```
Comment (samruddhikhandale) [inline test/java/scenarios.json:73]: So happy to see the added test coverage ✨

**Review by samruddhikhandale — CHANGES_REQUESTED** [review]
Another super helpful PR, thank you! 🙏 

Left some minor comments and questions!

File: src/java/install.sh (hunk 2 above)
Comment (jdputschadi) [reply src/java/install.sh:72]: Thanks, same as I did for the "git" feature, incorporated your feedback from the PR here.

File: src/java/install.sh (hunk 1 above)
Comment (jdputschadi) [reply src/java/install.sh]: Addressed in commit [85101db](https://github.com/devcontainers/features/pull/821/commits/85101db309e6bdbbd96dadf2484c3bde95a38006)

File: src/java/install.sh (hunk 4 above)
Comment (jdputschadi) [reply src/java/install.sh]: Addressed in commit [85101db](https://github.com/devcontainers/features/pull/821/commits/85101db309e6bdbbd96dadf2484c3bde95a38006)

**Comment by jdputschadi** [comment]
Samruddhi, commit 85101db addresses your suggestions, with all tests passing on my local system.

**Review by samruddhikhandale — CHANGES_REQUESTED** [review]
I have two open questions, and I'd let the tests run. Then we are good to go!

File: src/java/install.sh (hunk 5 above)
Comment (jdputschadi) [reply src/java/install.sh:245]: Some RHEL distributions do not install the "findutils", "util-linux", and "tar" packages by default. Mariner linux, for example does not install "util-linux" or "tar". Rocky linux minimal distributions do not install "findutils" or "tar" by default, and Alma linux minimal distributions do not install any of these packages by default.

These packages provide utilities that are needed to for "install.sh" to work.

File: src/java/install.sh (hunk 6 above)
Comment (jdputschadi) [reply src/java/install.sh:246]: Some RHEL distributions do not install the "findutils", "util-linux", and "tar" packages by default. Mariner linux, for example does not install "util-linux" or "tar". Rocky linux minimal distributions do not install "findutils" or "tar" by default, and Alma linux minimal distributions do not install any of these packages by default.

The same reasoning applies for "passwd" and "shadow-utils": they are required for "install.sh" to work.

These packages provide utilities that are needed to for "install.sh" to work.

**Comment by jdputschadi** [comment]
I've answered you're remaining two questions, but not resolved them. Please resolve the remaining questions if you find my answer satisfactory.

File: src/java/install.sh (hunk 5 above)
Comment (samruddhikhandale) [reply src/java/install.sh:245]: Makes sense, thanks for the explanation.

**Review by samruddhikhandale — APPROVED** [review]
Looks great, thank you for continuing to contribute these wonderful changes 👏

**Pull request merged**